File: blockchain/scripts/deploy_quantumagi.py
#!/usr/bin/env python3
"""
Quantumagi Deployment and Integration Script
Handles deployment of Quantumagi programs and integration with ACGS
"""

import asyncio
import json
import subprocess
import sys
import os
from pathlib import Path
from typing import Dict, List, Optional
import logging

# Setup logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class QuantumagiDeployer:
    """Handles deployment and integration of Quantumagi framework"""

    def __init__(self, config_path: str = "deploy_config.json"):
        self.config_path = config_path
        self.config = self._load_config()
        self.project_root = Path(__file__).parent.parent

    def _load_config(self) -> Dict:
        """Load deployment configuration"""
        default_config = {
            "solana_cluster": "devnet",
            "anchor_provider_url": "https://api.devnet.solana.com",
            "program_keypair_path": "target/deploy/quantumagi_core-keypair.json",
            "deployer_keypair_path": "~/.config/solana/id.json",
            "constitution_document_url": "https://arweave.net/constitution_hash",
            "gs_engine_config": {
                "llm_model": "gpt-4",
                "validation_threshold": 0.85,
                "max_policy_length": 1000,
            },
            "integration": {
                "acgs_backend_url": "http://localhost:8000",
                "enable_acgs_integration": True,
                "sync_policies": True,
            },
        }

        try:
            if os.path.exists(self.config_path):
                with open(self.config_path, "r") as f:
                    user_config = json.load(f)
                    default_config.update(user_config)
        except Exception as e:
            logger.warning(f"Could not load config from {self.config_path}: {e}")

        return default_config

    async def deploy_full_stack(self):
        """Deploy the complete Quantumagi stack"""
        logger.info("🚀 Starting Quantumagi full stack deployment")

        try:
            # Step 1: Build and deploy Solana program
            await self._build_and_deploy_program()

            # Step 2: Initialize constitution
            await self._initialize_constitution()

            # Step 3: Setup GS Engine
            await self._setup_gs_engine()

            # Step 4: Deploy initial policies
            await self._deploy_initial_policies()

            # Step 5: Integrate with ACGS backend
            if self.config["integration"]["enable_acgs_integration"]:
                await self._integrate_with_acgs()

            # Step 6: Run validation tests
            await self._run_validation_tests()

            logger.info("✅ Quantumagi deployment completed successfully!")
            await self._print_deployment_summary()

        except Exception as e:
            logger.error(f"❌ Deployment failed: {e}")
            raise

    async def _build_and_deploy_program(self):
        """Build and deploy the Solana program"""
        logger.info("📦 Building Solana program...")

        # Change to project directory
        os.chdir(self.project_root)

        # Build the program
        result = subprocess.run(["anchor", "build"], capture_output=True, text=True)
        if result.returncode != 0:
            raise Exception(f"Build failed: {result.stderr}")

        logger.info("✅ Program built successfully")

        # Deploy the program
        logger.info(f"🚀 Deploying to {self.config['solana_cluster']}...")

        deploy_cmd = [
            "anchor",
            "deploy",
            "--provider.cluster",
            self.config["solana_cluster"],
            "--provider.wallet",
            self.config["deployer_keypair_path"],
        ]

        result = subprocess.run(deploy_cmd, capture_output=True, text=True)
        if result.returncode != 0:
            raise Exception(f"Deployment failed: {result.stderr}")

        logger.info("✅ Program deployed successfully")

        # Extract program ID from deployment output
        program_id = self._extract_program_id(result.stdout)
        self.config["program_id"] = program_id
        logger.info(f"📋 Program ID: {program_id}")

    def _extract_program_id(self, deploy_output: str) -> str:
        """Extract program ID from anchor deploy output"""
        # Parse the deployment output to find program ID
        lines = deploy_output.split("\n")
        for line in lines:
            if "Program Id:" in line:
                return line.split(":")[-1].strip()

        # Fallback: read from keypair file
        try:
            with open(self.config["program_keypair_path"], "r") as f:
                keypair_data = json.load(f)
                # Convert keypair to program ID (simplified)
                return "Fg6PaFpoGXkYsidMpWTK6W2BeZ7FEfcYkg476zPFsLnS"  # Placeholder
        except:
            return "Fg6PaFpoGXkYsidMpWTK6W2BeZ7FEfcYkg476zPFsLnS"  # Default

    async def _initialize_constitution(self):
        """Initialize the constitutional governance system"""
        logger.info("🏛️ Initializing constitution...")

        # Create constitution hash
        constitution_content = await self._fetch_constitution_document()
        constitution_hash = self._create_constitution_hash(constitution_content)

        # Initialize using client
        from client.solana_client import QuantumagiSolanaClient
        from solders.keypair import Keypair

        # Load deployer keypair
        deployer_keypair = self._load_keypair(self.config["deployer_keypair_path"])

        client = QuantumagiSolanaClient(
            rpc_url=self.config["anchor_provider_url"],
            program_id=self.config["program_id"],
            payer_keypair=deployer_keypair,
        )

        signature = await client.initialize_constitution(constitution_hash)
        logger.info(f"✅ Constitution initialized. Signature: {signature}")

    async def _fetch_constitution_document(self) -> str:
        """Fetch the constitutional document"""
        # For demo, return a sample constitution
        return """
        Quantumagi Constitutional Framework v1.0
        
        Article I: Fundamental Principles
        1. No unauthorized state mutations (PC-001)
        2. Governance approval required for critical operations
        3. Transparency in all policy decisions
        
        Article II: AI Governance
        1. AI systems must operate within constitutional bounds
        2. Prompt governance compiler enforces real-time compliance
        3. Multi-model validation ensures policy reliability
        
        Article III: Democratic Governance
        1. Policy proposals require community voting
        2. Constitutional amendments require supermajority
        3. Emergency powers limited to critical situations
        """

    def _create_constitution_hash(self, content: str) -> bytes:
        """Create SHA-256 hash of constitution"""
        import hashlib

        return hashlib.sha256(content.encode()).digest()

    def _load_keypair(self, keypair_path: str) -> "Keypair":
        """Load Solana keypair from file"""
        from solders.keypair import Keypair

        # Expand user path
        path = os.path.expanduser(keypair_path)

        try:
            with open(path, "r") as f:
                keypair_data = json.load(f)
                return Keypair.from_bytes(bytes(keypair_data))
        except Exception as e:
            logger.warning(f"Could not load keypair from {path}: {e}")
            # Generate a new keypair for demo
            return Keypair()

    async def _setup_gs_engine(self):
        """Setup the Governance Synthesis Engine"""
        logger.info("🧠 Setting up GS Engine...")

        # Install Python dependencies
        gs_engine_dir = self.project_root / "gs_engine"
        requirements_file = gs_engine_dir / "requirements.txt"

        if requirements_file.exists():
            result = subprocess.run(
                [sys.executable, "-m", "pip", "install", "-r", str(requirements_file)],
                capture_output=True,
                text=True,
            )

            if result.returncode != 0:
                logger.warning(
                    f"Failed to install GS Engine dependencies: {result.stderr}"
                )

        # Test GS Engine
        try:
            from gs_engine.governance_synthesis import QuantumagiGSEngine

            gs_engine = QuantumagiGSEngine(self.config["gs_engine_config"])
            logger.info("✅ GS Engine initialized successfully")

        except Exception as e:
            logger.warning(f"GS Engine setup incomplete: {e}")

    async def _deploy_initial_policies(self):
        """Deploy initial governance policies"""
        logger.info("📜 Deploying initial policies...")

        initial_policies = [
            {
                "id": "PC-001",
                "title": "No Extrajudicial State Mutation",
                "content": "AI systems must not perform unauthorized state mutations without proper governance approval",
                "category": "prompt_constitution",
            },
            {
                "id": "SF-001",
                "title": "Safety Validation Required",
                "content": "All safety-critical operations must pass validation before execution",
                "category": "safety",
            },
            {
                "id": "GV-001",
                "title": "Governance Approval Threshold",
                "content": "Major governance decisions require 60% approval threshold",
                "category": "governance",
            },
        ]

        # Deploy each policy
        for policy_spec in initial_policies:
            try:
                await self._deploy_single_policy(policy_spec)
                logger.info(f"✅ Deployed policy: {policy_spec['id']}")
            except Exception as e:
                logger.warning(f"Failed to deploy policy {policy_spec['id']}: {e}")

    async def _deploy_single_policy(self, policy_spec: Dict):
        """Deploy a single policy"""
        from client.solana_client import QuantumagiSolanaClient
        from gs_engine.governance_synthesis import PolicyCategory

        # Map category string to enum
        category_map = {
            "prompt_constitution": PolicyCategory.PROMPT_CONSTITUTION,
            "safety": PolicyCategory.SAFETY,
            "governance": PolicyCategory.GOVERNANCE,
            "financial": PolicyCategory.FINANCIAL,
        }

        category = category_map.get(policy_spec["category"], PolicyCategory.GOVERNANCE)

        # Deploy using client
        deployer_keypair = self._load_keypair(self.config["deployer_keypair_path"])

        client = QuantumagiSolanaClient(
            rpc_url=self.config["anchor_provider_url"],
            program_id=self.config["program_id"],
            payer_keypair=deployer_keypair,
        )

        signature = await client.propose_policy_from_principle(
            policy_spec["id"], policy_spec["content"], category
        )

        # Auto-enact for initial deployment
        await asyncio.sleep(1)  # Wait for confirmation
        policy_id = hash(policy_spec["id"]) % 1000000  # Simple ID generation
        await client.enact_policy(policy_id)

    async def _integrate_with_acgs(self):
        """Integrate with existing ACGS backend"""
        logger.info("🔗 Integrating with ACGS backend...")

        try:
            # Test connection to ACGS backend
            import aiohttp

            async with aiohttp.ClientSession() as session:
                async with session.get(
                    f"{self.config['integration']['acgs_backend_url']}/health"
                ) as response:
                    if response.status == 200:
                        logger.info("✅ ACGS backend connection successful")
                    else:
                        logger.warning(
                            f"ACGS backend returned status: {response.status}"
                        )

        except Exception as e:
            logger.warning(f"Could not connect to ACGS backend: {e}")

    async def _run_validation_tests(self):
        """Run validation tests to ensure deployment success"""
        logger.info("🧪 Running validation tests...")

        try:
            # Run anchor tests
            result = subprocess.run(
                ["anchor", "test", "--skip-deploy"],
                capture_output=True,
                text=True,
                cwd=self.project_root,
            )

            if result.returncode == 0:
                logger.info("✅ All validation tests passed")
            else:
                logger.warning(f"Some tests failed: {result.stderr}")

        except Exception as e:
            logger.warning(f"Could not run validation tests: {e}")

    async def _print_deployment_summary(self):
        """Print deployment summary"""
        summary = f"""
        
🎉 Quantumagi Deployment Summary
================================

Program ID: {self.config.get('program_id', 'N/A')}
Cluster: {self.config['solana_cluster']}
Constitution: Initialized ✅
Initial Policies: Deployed ✅
GS Engine: Ready ✅

Next Steps:
1. Test compliance checking with: python -m client.solana_client
2. Integrate with your dApp using the client library
3. Monitor governance through the Solana explorer

Documentation: ./README.md
Client Examples: ./client/
        """

        print(summary)


async def main():
    """Main deployment function"""
    deployer = QuantumagiDeployer()

    try:
        await deployer.deploy_full_stack()
    except KeyboardInterrupt:
        logger.info("Deployment cancelled by user")
    except Exception as e:
        logger.error(f"Deployment failed: {e}")
        sys.exit(1)


if __name__ == "__main__":
    asyncio.run(main())

File: blockchain/scripts/deploy_quantumagi.rs
//! Quantumagi Deployment and Integration Tool - Rust Implementation
//! Handles deployment of Quantumagi programs and integration with ACGS
//! Replaces blockchain/scripts/deploy_quantumagi.py with native Rust implementation

use acgs_blockchain_client::{AcgsClient, governance::ConstitutionalPrinciple};
use anchor_client::solana_sdk::signature::Keypair;
use anyhow::{Result, Context};
use clap::{Parser, Subcommand};
use colored::*;
use serde::{Deserialize, Serialize};
use sha2::{Sha256, Digest};
use std::{
    collections::HashMap,
    fs,
    path::{Path, PathBuf},
    process::Command,
    time::{SystemTime, UNIX_EPOCH},
};

#[derive(Parser)]
#[command(name = "deploy_quantumagi")]
#[command(about = "Quantumagi Deployment and Integration Tool")]
struct Cli {
    #[command(subcommand)]
    command: Commands,
    
    /// Configuration file path
    #[arg(short, long, default_value = "deploy_config.json")]
    config: String,
    
    /// Verbose output
    #[arg(short, long)]
    verbose: bool,
}

#[derive(Subcommand)]
enum Commands {
    /// Deploy the complete Quantumagi stack
    Deploy,
    /// Build Solana programs only
    Build,
    /// Initialize constitution only
    InitConstitution,
    /// Deploy initial policies
    DeployPolicies,
    /// Run validation tests
    Validate,
    /// Show deployment status
    Status,
}

#[derive(Debug, Serialize, Deserialize)]
struct DeployConfig {
    solana_cluster: String,
    anchor_provider_url: String,
    program_keypair_path: String,
    deployer_keypair_path: String,
    constitution_document_url: String,
    gs_engine_config: GsEngineConfig,
    integration: IntegrationConfig,
    #[serde(skip_serializing_if = "Option::is_none")]
    program_ids: Option<HashMap<String, String>>,
}

#[derive(Debug, Serialize, Deserialize)]
struct GsEngineConfig {
    llm_model: String,
    validation_threshold: f64,
    max_policy_length: usize,
}

#[derive(Debug, Serialize, Deserialize)]
struct IntegrationConfig {
    acgs_backend_url: String,
    enable_acgs_integration: bool,
    sync_policies: bool,
}

#[derive(Debug, Serialize, Deserialize)]
struct PolicySpec {
    id: String,
    title: String,
    content: String,
    category: String,
}

pub struct QuantumagiDeployer {
    config: DeployConfig,
    project_root: PathBuf,
    verbose: bool,
}

impl QuantumagiDeployer {
    pub fn new(config_path: &str, verbose: bool) -> Result<Self> {
        let config = Self::load_config(config_path)?;
        let project_root = Path::new(env!("CARGO_MANIFEST_DIR")).parent()
            .context("Could not determine project root")?
            .to_path_buf();

        Ok(Self {
            config,
            project_root,
            verbose,
        })
    }

    fn load_config(config_path: &str) -> Result<DeployConfig> {
        let default_config = DeployConfig {
            solana_cluster: "devnet".to_string(),
            anchor_provider_url: "https://api.devnet.solana.com".to_string(),
            program_keypair_path: "target/deploy/quantumagi_core-keypair.json".to_string(),
            deployer_keypair_path: "~/.config/solana/id.json".to_string(),
            constitution_document_url: "https://arweave.net/constitution_hash".to_string(),
            gs_engine_config: GsEngineConfig {
                llm_model: "qwen/qwen3-32b".to_string(),
                validation_threshold: 0.85,
                max_policy_length: 1000,
            },
            integration: IntegrationConfig {
                acgs_backend_url: "http://localhost:8000".to_string(),
                enable_acgs_integration: true,
                sync_policies: true,
            },
            program_ids: None,
        };

        if Path::new(config_path).exists() {
            let config_content = fs::read_to_string(config_path)
                .context("Failed to read config file")?;
            let mut config: DeployConfig = serde_json::from_str(&config_content)
                .context("Failed to parse config file")?;
            
            // Merge with defaults for missing fields
            if config.program_ids.is_none() {
                config.program_ids = default_config.program_ids;
            }
            
            Ok(config)
        } else {
            println!("{} Config file not found, using defaults", "⚠️".yellow());
            Ok(default_config)
        }
    }

    pub async fn deploy_full_stack(&mut self) -> Result<()> {
        println!("{} Starting Quantumagi full stack deployment", "🚀".green());

        // Step 1: Build and deploy Solana programs
        self.build_and_deploy_programs().await?;

        // Step 2: Initialize constitution
        self.initialize_constitution().await?;

        // Step 3: Deploy initial policies
        self.deploy_initial_policies().await?;

        // Step 4: Integrate with ACGS backend
        if self.config.integration.enable_acgs_integration {
            self.integrate_with_acgs().await?;
        }

        // Step 5: Run validation tests
        self.run_validation_tests().await?;

        println!("{} Quantumagi deployment completed successfully!", "✅".green());
        self.print_deployment_summary().await?;

        Ok(())
    }

    pub async fn build_and_deploy_programs(&mut self) -> Result<()> {
        println!("{} Building Solana programs...", "📦".blue());

        // Change to project directory
        std::env::set_current_dir(&self.project_root)?;

        // Build the programs
        let build_output = Command::new("anchor")
            .args(["build"])
            .output()
            .context("Failed to execute anchor build")?;

        if !build_output.status.success() {
            let error = String::from_utf8_lossy(&build_output.stderr);
            anyhow::bail!("Build failed: {}", error);
        }

        println!("{} Programs built successfully", "✅".green());

        // Deploy the programs
        println!("{} Deploying to {}...", "🚀".blue(), self.config.solana_cluster);

        let deploy_output = Command::new("anchor")
            .args([
                "deploy",
                "--provider.cluster", &self.config.solana_cluster,
                "--provider.wallet", &self.config.deployer_keypair_path,
            ])
            .output()
            .context("Failed to execute anchor deploy")?;

        if !deploy_output.status.success() {
            let error = String::from_utf8_lossy(&deploy_output.stderr);
            anyhow::bail!("Deployment failed: {}", error);
        }

        println!("{} Programs deployed successfully", "✅".green());

        // Extract and store program IDs
        let deploy_stdout = String::from_utf8_lossy(&deploy_output.stdout);
        self.extract_and_store_program_ids(&deploy_stdout)?;

        Ok(())
    }

    fn extract_and_store_program_ids(&mut self, deploy_output: &str) -> Result<()> {
        let mut program_ids = HashMap::new();

        // Parse deployment output for program IDs
        for line in deploy_output.lines() {
            if line.contains("Program Id:") {
                let parts: Vec<&str> = line.split(':').collect();
                if parts.len() >= 2 {
                    let program_id = parts[1].trim();
                    
                    // Determine program name based on context
                    if line.contains("quantumagi_core") || line.contains("governance") {
                        program_ids.insert("quantumagi_core".to_string(), program_id.to_string());
                    } else if line.contains("appeals") {
                        program_ids.insert("appeals".to_string(), program_id.to_string());
                    } else if line.contains("logging") {
                        program_ids.insert("logging".to_string(), program_id.to_string());
                    }
                }
            }
        }

        // Fallback: read from devnet_program_ids.json if available
        let devnet_ids_path = self.project_root.join("devnet_program_ids.json");
        if program_ids.is_empty() && devnet_ids_path.exists() {
            let devnet_content = fs::read_to_string(&devnet_ids_path)?;
            let devnet_data: serde_json::Value = serde_json::from_str(&devnet_content)?;
            
            if let Some(programs) = devnet_data.get("programs").and_then(|p| p.as_object()) {
                for (name, id) in programs {
                    if let Some(id_str) = id.as_str() {
                        program_ids.insert(name.clone(), id_str.to_string());
                    }
                }
            }
        }

        self.config.program_ids = Some(program_ids.clone());

        // Print program IDs
        for (name, id) in &program_ids {
            println!("{} {}: {}", "📋".blue(), name, id);
        }

        Ok(())
    }

    pub async fn initialize_constitution(&self) -> Result<()> {
        println!("{} Initializing constitution...", "🏛️".blue());

        // Create constitution content
        let constitution_content = self.fetch_constitution_document().await?;
        let constitution_hash = self.create_constitution_hash(&constitution_content);

        // Load deployer keypair
        let deployer_keypair = self.load_keypair(&self.config.deployer_keypair_path)?;

        // Create client
        let client = AcgsClient::devnet(deployer_keypair)?;

        // Create constitutional principles
        let principles = self.get_constitutional_principles();

        // Initialize constitution
        let signature = client.initialize_constitution(principles).await?;
        
        println!("{} Constitution initialized. Signature: {}", "✅".green(), signature);

        Ok(())
    }

    async fn fetch_constitution_document(&self) -> Result<String> {
        // For now, return embedded constitution content
        Ok(r#"
        Quantumagi Constitutional Framework v1.0

        Article I: Fundamental Principles
        1. No unauthorized state mutations (PC-001)
        2. Governance approval required for critical operations
        3. Transparency in all policy decisions

        Article II: AI Governance
        1. AI systems must operate within constitutional bounds
        2. Prompt governance compiler enforces real-time compliance
        3. Multi-model validation ensures policy reliability

        Article III: Democratic Governance
        1. Policy proposals require community voting
        2. Constitutional amendments require supermajority
        3. Emergency powers limited to critical situations
        "#.to_string())
    }

    fn create_constitution_hash(&self, content: &str) -> [u8; 32] {
        let mut hasher = Sha256::new();
        hasher.update(content.as_bytes());
        hasher.finalize().into()
    }

    fn get_constitutional_principles(&self) -> Vec<ConstitutionalPrinciple> {
        let timestamp = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs() as i64;

        vec![
            ConstitutionalPrinciple {
                id: 1,
                title: "No Extrajudicial State Mutation".to_string(),
                content: "AI systems must not perform unauthorized state mutations without proper governance approval".to_string(),
                is_active: true,
                created_at: timestamp,
            },
            ConstitutionalPrinciple {
                id: 2,
                title: "Safety Validation Required".to_string(),
                content: "All safety-critical operations must pass validation before execution".to_string(),
                is_active: true,
                created_at: timestamp,
            },
            ConstitutionalPrinciple {
                id: 3,
                title: "Governance Approval Threshold".to_string(),
                content: "Major governance decisions require 60% approval threshold".to_string(),
                is_active: true,
                created_at: timestamp,
            },
        ]
    }

    fn load_keypair(&self, keypair_path: &str) -> Result<Keypair> {
        let path = if keypair_path.starts_with('~') {
            keypair_path.replace('~', &std::env::var("HOME").unwrap_or_default())
        } else {
            keypair_path.to_string()
        };

        let keypair_data = fs::read_to_string(&path)
            .context("Failed to read keypair file")?;

        let keypair_bytes: Vec<u8> = serde_json::from_str(&keypair_data)
            .context("Failed to parse keypair JSON")?;

        Ok(Keypair::from_bytes(&keypair_bytes)?)
    }

    pub async fn deploy_initial_policies(&self) -> Result<()> {
        println!("{} Deploying initial policies...", "📜".blue());

        let initial_policies = self.get_initial_policies();

        // Load deployer keypair
        let deployer_keypair = self.load_keypair(&self.config.deployer_keypair_path)?;
        let client = AcgsClient::devnet(deployer_keypair)?;

        // Deploy each policy
        for (i, policy_spec) in initial_policies.iter().enumerate() {
            match self.deploy_single_policy(&client, policy_spec).await {
                Ok(_) => {
                    println!("{} Deployed policy: {}", "✅".green(), policy_spec.id);
                }
                Err(e) => {
                    println!("{} Failed to deploy policy {}: {}", "⚠️".yellow(), policy_spec.id, e);
                }
            }

            // Small delay between deployments
            tokio::time::sleep(tokio::time::Duration::from_millis(500)).await;
        }

        Ok(())
    }

    fn get_initial_policies(&self) -> Vec<PolicySpec> {
        vec![
            PolicySpec {
                id: "PC-001".to_string(),
                title: "No Extrajudicial State Mutation".to_string(),
                content: "AI systems must not perform unauthorized state mutations without proper governance approval".to_string(),
                category: "prompt_constitution".to_string(),
            },
            PolicySpec {
                id: "SF-001".to_string(),
                title: "Safety Validation Required".to_string(),
                content: "All safety-critical operations must pass validation before execution".to_string(),
                category: "safety".to_string(),
            },
            PolicySpec {
                id: "GV-001".to_string(),
                title: "Governance Approval Threshold".to_string(),
                content: "Major governance decisions require 60% approval threshold".to_string(),
                category: "governance".to_string(),
            },
        ]
    }

    async fn deploy_single_policy(&self, client: &AcgsClient, policy_spec: &PolicySpec) -> Result<()> {
        // Submit proposal
        let signature = client.submit_proposal(
            policy_spec.title.clone(),
            policy_spec.content.clone(),
        ).await?;

        if self.verbose {
            println!("  Policy proposal signature: {}", signature);
        }

        // Auto-approve for initial deployment (in production, this would require voting)
        tokio::time::sleep(tokio::time::Duration::from_millis(1000)).await;

        // Generate a simple policy ID for voting
        let policy_id = (policy_spec.id.chars().map(|c| c as u64).sum::<u64>()) % 1000000;

        let vote_signature = client.vote_on_proposal(policy_id, true).await?;

        if self.verbose {
            println!("  Policy vote signature: {}", vote_signature);
        }

        Ok(())
    }

    pub async fn integrate_with_acgs(&self) -> Result<()> {
        println!("{} Integrating with ACGS backend...", "🔗".blue());

        let client = reqwest::Client::new();
        let health_url = format!("{}/health", self.config.integration.acgs_backend_url);

        match client.get(&health_url).send().await {
            Ok(response) => {
                if response.status().is_success() {
                    println!("{} ACGS backend connection successful", "✅".green());
                } else {
                    println!("{} ACGS backend returned status: {}", "⚠️".yellow(), response.status());
                }
            }
            Err(e) => {
                println!("{} Could not connect to ACGS backend: {}", "⚠️".yellow(), e);
            }
        }

        Ok(())
    }

    pub async fn run_validation_tests(&self) -> Result<()> {
        println!("{} Running validation tests...", "🧪".blue());

        // Change to project directory
        std::env::set_current_dir(&self.project_root)?;

        // Run Rust tests
        let rust_test_output = Command::new("cargo")
            .args(["test", "--manifest-path", "client/rust/Cargo.toml", "--", "--nocapture"])
            .output();

        match rust_test_output {
            Ok(output) => {
                if output.status.success() {
                    println!("{} Rust validation tests passed", "✅".green());
                } else {
                    let stderr = String::from_utf8_lossy(&output.stderr);
                    println!("{} Some Rust tests failed: {}", "⚠️".yellow(), stderr);
                }
            }
            Err(e) => {
                println!("{} Could not run Rust validation tests: {}", "⚠️".yellow(), e);
            }
        }

        // Run anchor tests if available
        let anchor_test_output = Command::new("anchor")
            .args(["test", "--skip-deploy"])
            .output();

        match anchor_test_output {
            Ok(output) => {
                if output.status.success() {
                    println!("{} Anchor validation tests passed", "✅".green());
                } else {
                    let stderr = String::from_utf8_lossy(&output.stderr);
                    println!("{} Some Anchor tests failed: {}", "⚠️".yellow(), stderr);
                }
            }
            Err(e) => {
                println!("{} Could not run Anchor validation tests: {}", "⚠️".yellow(), e);
            }
        }

        Ok(())
    }

    pub async fn print_deployment_summary(&self) -> Result<()> {
        let program_ids = self.config.program_ids.as_ref()
            .map(|ids| {
                ids.iter()
                    .map(|(name, id)| format!("  {}: {}", name, id))
                    .collect::<Vec<_>>()
                    .join("\n")
            })
            .unwrap_or_else(|| "  N/A".to_string());

        let summary = format!(r#"

{} Quantumagi Deployment Summary
================================

Program IDs:
{}
Cluster: {}
Constitution: Initialized ✅
Initial Policies: Deployed ✅
ACGS Integration: {} ✅

Next Steps:
1. Test compliance checking with: cargo run --bin deploy_quantumagi validate
2. Integrate with your dApp using the Rust client library
3. Monitor governance through the Solana explorer

Documentation: ./README.md
Client Library: ./client/rust/
        "#,
            "🎉".green(),
            program_ids,
            self.config.solana_cluster,
            if self.config.integration.enable_acgs_integration { "Enabled" } else { "Disabled" }
        );

        println!("{}", summary);

        Ok(())
    }

    pub async fn show_status(&self) -> Result<()> {
        println!("{} Quantumagi Deployment Status", "📊".blue());

        if let Some(program_ids) = &self.config.program_ids {
            println!("\n{} Deployed Programs:", "📋".blue());
            for (name, id) in program_ids {
                println!("  {}: {}", name, id);
            }
        } else {
            println!("{} No programs deployed yet", "⚠️".yellow());
        }

        println!("\n{} Configuration:", "⚙️".blue());
        println!("  Cluster: {}", self.config.solana_cluster);
        println!("  RPC URL: {}", self.config.anchor_provider_url);
        println!("  ACGS Integration: {}",
                 if self.config.integration.enable_acgs_integration { "Enabled" } else { "Disabled" });

        Ok(())
    }
}

#[tokio::main]
async fn main() -> Result<()> {
    let cli = Cli::parse();

    let mut deployer = QuantumagiDeployer::new(&cli.config, cli.verbose)?;

    match cli.command {
        Commands::Deploy => {
            deployer.deploy_full_stack().await?;
        }
        Commands::Build => {
            deployer.build_and_deploy_programs().await?;
        }
        Commands::InitConstitution => {
            deployer.initialize_constitution().await?;
        }
        Commands::DeployPolicies => {
            deployer.deploy_initial_policies().await?;
        }
        Commands::Validate => {
            deployer.run_validation_tests().await?;
        }
        Commands::Status => {
            deployer.show_status().await?;
        }
    }

    Ok(())
}
